modules/azure/client_factory.go

```diff
@@ -0,0 +1,147 @@
+/*
+
+This file implements an Azure client factory that automatically handles setting up Base URI
+values for sovereign cloud support. Note the list of clients below is not initially exhaustive;
+rather, additional clients will me added as-needed.
+
+*/
+
+package azure
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-07-01/compute"
+	"github.com/Azure/azure-sdk-for-go/services/containerservice/mgmt/2019-11-01/containerservice"
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-06-01/subscriptions"
+	autorestAzure "github.com/Azure/go-autorest/autorest/azure"
+)
+
+const (
+	// AzureEnvironmentEnvName is the name of the Azure environment to use. Set to one of the following:
+	//
+	// "AzureChinaCloud":        ChinaCloud
+	// "AzureGermanCloud":       GermanCloud
+	// "AzurePublicCloud":       PublicCloud
+	// "AzureUSGovernmentCloud": USGovernmentCloud
+	// "AzureStackCloud":		 Azure stack
+	AzureEnvironmentEnvName = "AZURE_ENVIRONMENT"
+)
+
+// ClientType describes the type of client a module can create.
+type ClientType int
+
+const (
+	// SubscriptionsClientType represents a SubscriptionClient
+	SubscriptionsClientType ClientType = iota
+
+	// VirtualMachinesClientType represents a VirtualMachinesClient
+	VirtualMachinesClientType
+
+	// ManagedClustersClientType represents a ManagedClustersClient
+	ManagedClustersClientType
+)
+
+// ClientFactory describes the methods available on client factory implementatoins
+type ClientFactory interface {
+	// GetClientE returns a client instance based on the ClientType passed, or optionally an error.
+	GetClientE(clientType ClientType, subscriptionID string) (interface{}, error)
+
+	// GetClientBaseURIE returns the configured BaseURI for the given client, based on the passed ClientType.
+	GetClientBaseURIE(clientType ClientType, client interface{}) (string, error)
+}
+
+// multiEnvClientFactory is used to coordinate handing out properly configured Azure SDK clients
+// that are properly setup for use with Public or Sovereign clouds (depending on configuration)
+type multiEnvClientFactory struct{}
```

Comment:
Why attach these methods to a `struct` rather than just have them as methods on the package? Using the `struct` requires an extra call to `NewClientFactory()` every time, and as the `struct` stores no data at all, it's not clear what the benefit of that is?

Reply:
The idea originally was to "hide" the inner workings of the client_factory and provide only the needed methods (e.g., "pit of success") as this wasn't intended to be used outside of modules. 

That said - we have removed the extra interface / struct and just made all methods top-level.